medmodels/data_synthesis/synthesizer_model.py

```diff
@@ -0,0 +1,84 @@
+"""This module defines the SynthesizerModel class, which is a PyTorch module that
+is used to generate synthetic data using a synthesizer model, after the synthesizer
+model has been trained on real data."""
+
+from abc import ABCMeta, abstractmethod
+from pathlib import Path
+from typing import Any, Dict, Union
+
+import sparse
+import torch
+from torch import nn
+
+from medmodels import MedRecord
+from medmodels.data_synthesis.synthesizer import Synthesizer
+
+
+class SynthesizerModel(nn.Module, metaclass=ABCMeta):
+    """SynthesizerModel is an abstract class that serves as a blueprint for synthesizer models.
+
+    This class is instantiated once a Synthesizer has been trained on real data. It is
+    separated from the Synthesizer class to allow for the generation of synthetic data
+    without incurring into side effects."""
+
+    number_of_samples: int
+    device: torch.device
+
+    def __init__(
+        self,
+        synthesizer: Synthesizer,
+        number_of_samples: int,
+        device: torch.device,
+        **kwargs: Dict[str, Any],
+    ) -> None:
+        """Initializes the SynthesizerModel with the required components.
+
+        Args:
+            synthesizer (Synthesizer): A trained synthesizer model.
+            number_of_samples (int): Number of samples to generate.
+            device (torch.device): Device to use for the model.
+            **kwargs (Dict[str, Any]): Additional keyword arguments.
+        """
+        super(SynthesizerModel, self).__init__()
+        self.number_of_samples = number_of_samples
+        self.device = device
+        self.to(self.device)
+
+    @abstractmethod
+    def forward(self) -> Union[torch.Tensor, sparse.COO]:
+        """Generates synthetic data using the trained synthesizer model.
+
+        Returns:
+            Union[torch.Tensor, sparse.COO]: Synthetic data generated by the model.
+        """
+        pass
+
+    @abstractmethod
+    def save_model(self, path: Path) -> None:
+        """Saves the trained synthesizer model to a file.
+
+        Args:
+            path (Path): Path to save the model to.
+        """
+        pass
+
+    @abstractmethod
+    def generate_synthetic_data(self) -> MedRecord:
+        """Generates synthetic data using the trained synthesizer model.#
```

Comment:
remove `#`

also go into detail that this MedRecord is not postprocessed yet and cannot be really used